torch_audiomentations/augmentations/impulse_response.py

```diff
@@ -0,0 +1,27 @@
+import numpy as np
+import random
+import torch
+from ..core.transforms_interface import BasicTransform
+from ..utils.convolution import convolve
+from ..utils.file import find_audio_files, load_audio
+
+
+class ApplyImpulseResponse(BasicTransform):
+    """
+    Convolves an input signal with an impulse response.
+    """
+
+    def __init__(self, ir_path, convolve_mode="full", p=0.5):
+        super(ApplyImpulseResponse, self).__init__(p)
+        self.ir_path = find_audio_files(ir_path)
+        self.convolve_mode = convolve_mode
+
+    def randomize_parameters(self, samples, sample_rate):
+        super(ApplyImpulseResponse, self).randomize_parameters(samples, sample_rate)
+        if self.parameters["should_apply"]:
+            self.parameters["ir_file_path"] = random.choice(self.ir_path)
```

Comment:
If we're going to support augmenting batches of sounds, we might want to use different random settings for each sound, right? In that case, I think we're going to need a vector of IR file paths here, and not just a single one

Reply:
https://github.com/asteroid-team/torch-audiomentations/issues/7

Reply:
This makes me realize that you plan to apply augmentations to batches of sounds, instead of individual sounds. Is this really what we want? In this case, do you handle the probabilistic side of augmentations per-batch or per-example?

Reply:
Yes, I think that makes sense performance-wise. [The plot](https://github.com/asteroid-team/torch-audiomentations/blob/master/images/convolve_exec_time_plot.png) where I crudely measured batched augmentation vs individual sounds (AKA batch size = 1) showed a 33% reduction in computation time. Further gains are possible with better GPUs.

Reply:
Do you still observe the performance gain if the IRs are different? 
I agree that this _should_ be a possibility, for performance reasons. But it should be possible to apply independent augmentations in the Dataset (which is done in parallel anyway)

Reply:
Right. Many people use multiple workers for the dataset generator. How well is GPU supported in such a multi-worker setup? Will it cause problems?

https://pytorch.org/docs/stable/notes/multiprocessing.html#cuda-in-multiprocessing

Reply:
> Do you still observe the performance gain if the IRs are different?

Good question! Indeed the performance gain is smaller when the IRs are different. Crudely measured:

```
torch FFT CUDA       batch size = 1,  single kernel   0.151 s
torch FFT CUDA       batch size = 32, single kernel   0.102 s
torch FFT CUDA       batch size = 32, 32 kernels      0.133 s
```

Reply:
And I don't know how well multiprocessing will behave on GPU.. That would be something to try before making a design choice. Have you looked a albumentations to see how do they do it?

Reply:
AFAIK, albumentations is CPU-only. This one seems to support GPU though: https://github.com/kornia/kornia

Reply:
Ok. 
And kornia's augmentations are done out of the dataset in the example. 
https://kornia.readthedocs.io/en/latest/tutorials/data_augmentation.html

Reply:
So should we support batches of IRs too? Or we haven't fully decided? Could this be possible as an improvement later on?

Reply:
Yes. Sure, we can do it later

Reply:
Re kornia's example: Concidentally, that was the way I thought of doing it in my training pipeline. I.e. let the data loader take care of loading data (using CPU only), then do a data augmentation step (on the GPU), then pass to the model. Thanks for verifying that this is a valid approach 😄